### PR TITLE
Fix dynamics grid IO

### PR DESCRIPTION
**Pull request by bartgol**
As reported in #1602, the heroic debugging of @wlin7 exposed a bug in the dyn grid IO, which required some adjustment. This PR fixes the bug, adds a unit test for dyn grid output, and expands the model restart test to run also with more than 1 mpi rank (if available).

### CODE REVIEW

**Comment by wlin7** [comment]
Wow. What a quick work! You are my hero, @bartgol . I cant even follow what some of the changes are for. I am going to run some tests first with your branch.

**Comment by wlin7** [comment]
ne4 test run failed. e3sm.log showed MPI abort called just after atm initialization. No other useful error log.
 ```
0: Create Pool
 0: [EAMXX] initialize_atm_procs ... done!
 0: Atmosphere step = 0
 0:   model time = 0001-01-01 00:00:00
 0:
 0: Rank 0 [Thu May 12 20:43:15 2022] [c1-1c1s14n3] application called MPI_Abort(MPI_COMM_WORLD, 128) - process 0
```

**Comment by bartgol** [comment]
Yes, I haven't ran any test yet. I was hoping to have the AT results this morning, but unfortunately mappy is always busy at night, and didn't make it through. I'll manually run something this morning, and see if I can track the errors down.

**Comment by bartgol** [comment]
The autotester is currently down. I have tested manually the branch on the tgt machines, and it passes on all of them.

@wlin7 I'm not sure why the test fails for you. Is this one of our v1 testsuite tests? Or were you running something particular?

**Comment by wlin7** [comment]
Thanks for the update, @bartgol . My failed test was a manual one, not out of the test suite. Encouraged by your successful test, I did a standard ERS test and they work well. The restart file has proper global dimension size for the dynamics grid (elem=96 instead of 4 before the fix) when running with 24 MPI task. The ERS test is BFB in terms of cpl.hi. All history output as specified in sream_output.yaml are also BFB in restart run. Note that in this ERS test, scream_output.yaml did not specify the three fields that showed slight difference in the restart run reported in [#1602](https://github.com/E3SM-Project/scream/pull/1602)).

I must have messed up something in my manual test. I will figure that out separately. It does not matter now.

File: components/scream/src/share/io/scorpio_input.cpp [hunk 1]
```diff
@@ -521,6 +521,36 @@ get_var_dof_offsets(const FieldLayout& layout)
       auto offset = (gid-min_gid)*col_size;
       std::iota(start,end,offset);
     }
+  } else if (layout.has_tag(ShortFieldTagsNames::EL)) {
+    auto layout2d = m_io_grid->get_2d_scalar_layout();
+    const int num_my_elems = layout2d.dim(0);
+    const int ngp = layout2d.dim(1);
+    const int num_cols = num_my_elems*ngp*ngp;
+
+    // Note: col_size might be *larger* than the number of vertical levels, or even smalle.
```
Comment (AaronDonahue) [inline components/scream/src/share/io/scorpio_input.cpp:530]: *"smaller"

File: components/scream/src/share/io/scorpio_output.cpp [hunk 2]
```diff
@@ -507,6 +507,36 @@ std::vector<int> AtmosphereOutput::get_var_dof_offsets(const FieldLayout& layout
       auto offset = (gid-min_gid)*col_size;
       std::iota(start,end,offset);
     }
+  } else if (layout.has_tag(ShortFieldTagsNames::EL)) {
+    auto layout2d = m_io_grid->get_2d_scalar_layout();
+    const int num_my_elems = layout2d.dim(0);
+    const int ngp = layout2d.dim(1);
+    const int num_cols = num_my_elems*ngp*ngp;
+
+    // Note: col_size might be *larger* than the number of vertical levels, or even smalle.
```
Comment (AaronDonahue) [inline components/scream/src/share/io/scorpio_output.cpp:516]: *"smaller"

**Review by AaronDonahue — APPROVED** [review]
Just two minor comments that don't need changing.  I leave it up to you @bartgol if you want to fix them or just go ahead and merge.

**Comment by bartgol** [comment]
> Just two minor comments that don't need changing. I leave it up to you @bartgol if you want to fix them or just go ahead and merge.

@AaronDonahue done!

**Comment by AaronDonahue** [comment]
Okay doubly approved and passing tests (confirmed by Luca) - so I am going to manually merge.

**Pull request merged**